### DISCUSSION TITLE
nwbtest() fails in Matlab

### DISCUSSION THREAD

**Discussion by laurelrr** [NeurodataWithoutBorders/helpdesk#9] [General]
From my nwb directory, I can generate the API with no issues but after running the test suite, a number of tests fail.  They seem to relate to python files however so it is not clear if I should be worried by this.  

MATLAB Version: 9.7.0.1737446 (R2019b) Update 9
Operating System: Microsoft Windows 11 Pro Insider Preview Version 10.0 (Build 22471)

>> addpath(genpath(pwd));
>> generateCore();
>> nwbtest()
....
================================================================================
Verification failed in tests.system.ElectricalSeriesIOTest/testOutToPyNWB.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testOutToPyNWB) at 15
================================================================================
.
================================================================================
Assertion failed in tests.system.ElectricalSeriesIOTest/testInFromPyNWB and it did not run to completion.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testInFromPyNWB) at 22
================================================================================
..
================================================================================
Verification failed in tests.system.ElectrodeGroupIOTest/testOutToPyNWB.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testOutToPyNWB) at 15
================================================================================
.
================================================================================
Assertion failed in tests.system.ElectrodeGroupIOTest/testInFromPyNWB and it did not run to completion.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testInFromPyNWB) at 22
================================================================================
.. 
================================================================================
Verification failed in tests.system.ImagingPlaneIOTest/testOutToPyNWB.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testOutToPyNWB) at 15
================================================================================
.
================================================================================
Assertion failed in tests.system.ImagingPlaneIOTest/testInFromPyNWB and it did not run to completion.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testInFromPyNWB) at 22
================================================================================
..
================================================================================
Verification failed in tests.system.NWBFileIOTest/testOutToPyNWB.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testOutToPyNWB) at 15
================================================================================
.
================================================================================
Assertion failed in tests.system.NWBFileIOTest/testInFromPyNWB and it did not run to completion.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testInFromPyNWB) at 22
================================================================================
...
================================================================================
Verification failed in tests.system.PhotonSeriesIOTest/testOutToPyNWB.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testOutToPyNWB) at 15
================================================================================
.
================================================================================
Assertion failed in tests.system.PhotonSeriesIOTest/testInFromPyNWB and it did not run to completion.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testInFromPyNWB) at 22
================================================================================
.. 
================================================================================
Verification failed in tests.system.TimeSeriesIOTest/testOutToPyNWB.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testOutToPyNWB) at 15
================================================================================
.
================================================================================
Assertion failed in tests.system.TimeSeriesIOTest/testInFromPyNWB and it did not run to completion.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 7, in <module>
        import h5py
    ModuleNotFoundError: No module named 'h5py'
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testInFromPyNWB) at 22
================================================================================
......... ......
Failure Summary:

     Name                                                 Failed  Incomplete  Reason(s)
    ==================================================================================================
     tests.system.ElectricalSeriesIOTest/testOutToPyNWB     X                 Failed by verification.
    --------------------------------------------------------------------------------------------------
     tests.system.ElectricalSeriesIOTest/testInFromPyNWB    X         X       Failed by assertion.
    --------------------------------------------------------------------------------------------------
     tests.system.ElectrodeGroupIOTest/testOutToPyNWB       X                 Failed by verification.
    --------------------------------------------------------------------------------------------------
     tests.system.ElectrodeGroupIOTest/testInFromPyNWB      X         X       Failed by assertion.
    --------------------------------------------------------------------------------------------------
     tests.system.ImagingPlaneIOTest/testOutToPyNWB         X                 Failed by verification.
    --------------------------------------------------------------------------------------------------
     tests.system.ImagingPlaneIOTest/testInFromPyNWB        X         X       Failed by assertion.
    --------------------------------------------------------------------------------------------------
     tests.system.NWBFileIOTest/testOutToPyNWB              X                 Failed by verification.
    --------------------------------------------------------------------------------------------------
     tests.system.NWBFileIOTest/testInFromPyNWB             X         X       Failed by assertion.
    --------------------------------------------------------------------------------------------------
     tests.system.PhotonSeriesIOTest/testOutToPyNWB         X                 Failed by verification.
    --------------------------------------------------------------------------------------------------
     tests.system.PhotonSeriesIOTest/testInFromPyNWB        X         X       Failed by assertion.
    --------------------------------------------------------------------------------------------------
     tests.system.TimeSeriesIOTest/testOutToPyNWB           X                 Failed by verification.
    --------------------------------------------------------------------------------------------------
     tests.system.TimeSeriesIOTest/testInFromPyNWB          X         X       Failed by assertion.

results = 

  1×36 TestResult array with properties:

    Name
    Passed
    Failed
    Incomplete
    Duration
    Details

Totals:
   24 Passed, 12 Failed (rerun), 6 Incomplete.
   164.855 seconds testing time.

ans = 

  1×36 TestResult array with properties:

    Name
    Passed
    Failed
    Incomplete
    Duration
    Details

Totals:
   24 Passed, 12 Failed (rerun), 6 Incomplete.
   164.855 seconds testing time.

**Comment by oruebel** [comment]
It looks like ``h5py`` (and maybe ``PyNWB``) may not be installed in your Python environment. MatNWB should not need those but they are required for round-trip-tests where MatNWB tests that it can read files from PyNWB and vice versa. Instructions for installing PyNWB are here https://pynwb.readthedocs.io/en/stable/getting_started.html#installation

**Reply by laurelrr** [reply]
I do have both matNWB and pyNWB installed on my machine.

**Reply by laurelrr** [reply]
Is there something I am supposed to do in Matlab to point to the pyNWB installation?

**Reply by bendichter** [reply]
Your computer likely has multiple versions of Python installed, and it's possible that the tests are pointing to a Python that does not have PyNWB installed. Maybe @ln-vidrio can help with this

**Reply by laurelrr** [reply]
Yes, that is possible.  When I first installed pynwb, I set up a conda environment for pynwb and installed everything there.  I just tried opening Matlab from the Anaconda prompt with pynwb activated and reran nwbtest() but got the same errors as before.

**Reply by bendichter** [reply]
Yeah I have also had trouble setting this up. If it's only the Python round trip tests that are failing I'd say MatNWB is working fine, and the failure of these tests does not mean that PyNWB is mis-installed, just that MatNWB and PyNWB aren't communicating properly. This is only needed for the tests, so unless you are planning on developing MatNWB itself, debugging the remaining tests may not be that important for you.

**Reply by lawrence-mbf** [reply]
I don't think MATLAB inherits the path from the shell unfortunately. The way I've worked around it is to optionally use a special configuration `env.mat` file saved in the `+tests` directory in matnwb. In the `env.mat` file should be a variable named `pythonDir` which is a string containing the path to your python environment. In my case, I'm using miniconda so my `pythonDir` is something like `...\Miniconda3\envs\pynwb` (it should be the full path).

**Reply by lawrence-mbf** [reply]
Note that this is special behavior I added in, nothing special about `env.mat` otherwise.

**Reply by laurelrr** [reply]
OK, I tried this and it didn't fix the issue for me.  Just to verify, here were my steps:
1. open anaconda prompt
2. $ conda info --envs
3. copy the filepath for pynwb to a matlab variable called pythonDir and save it to env.mat in the path_to_nwb\+tests\ directory.
4. restart matlab
5. cd to nwb directory; addpath(genpath(pwd)); generateCore();
6. nwbtests()

It still errors on the same files as before but with some additional info included in the error.  I include the firs failed test here:
```
================================================================================
Verification failed in tests.system.ElectricalSeriesIOTest/testOutToPyNWB.
    ----------------
    Test Diagnostic:
    ----------------
    E
    ======================================================================
    ERROR: PyNWBIOTest (unittest.loader._FailedTest)
    ----------------------------------------------------------------------
    ImportError: Failed to import test module: PyNWBIOTest
    Traceback (most recent call last):
      File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\numpy\core\__init__.py", line 22, in <module>
        from . import multiarray
      File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\numpy\core\multiarray.py", line 12, in <module>
        from . import overrides
      File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\numpy\core\overrides.py", line 7, in <module>
        from numpy.core._multiarray_umath import (
    ImportError: DLL load failed while importing _multiarray_umath: The specified module could not be found.
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\unittest\loader.py", line 154, in loadTestsFromName
        module = __import__(module_name)
      File "C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.py", line 4, in <module>
        import numpy as np
      File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\numpy\__init__.py", line 150, in <module>
        from . import core
      File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\numpy\core\__init__.py", line 48, in <module>
        raise ImportError(msg)
    ImportError: 
    
    IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!
    
    Importing the numpy C-extensions failed. This error can happen for
    many reasons, often due to issues with your setup or how NumPy was
    installed.
    
    We have compiled some common reasons and troubleshooting tips at:
    
        https://numpy.org/devdocs/user/troubleshooting-importerror.html
    
    Please note and check the following:
    
      * The Python version is: Python3.8 from "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\python.exe"
      * The NumPy version is: "1.21.2"
    
    and make sure that they are the versions you expect.
    Please carefully study the documentation linked above for further help.
    
    Original error was: DLL load failed while importing _multiarray_umath: The specified module could not be found.
    
    
    
    ----------------------------------------------------------------------
    Ran 1 test in 0.000s
    
    FAILED (errors=1)
    
    ------------------
    Stack Information:
    ------------------
    In C:\Users\Laurel\Documents\MATLAB\nwb\+tests\+system\PyNWBIOTest.m (PyNWBIOTest.testOutToPyNWB) at 15
================================================================================
.
```

**Reply by lawrence-mbf** [reply]
That's very odd. The error suggests that this is an issue with the numpy installation. Are you sure that the pynwb installation works as expected?

**Reply by laurelrr** [reply]
Well, thank you for the responses.  I am OK with this for the time being.  I don't expect to develop MatNWB at the moment. I have not used the pynwb installation extensively but it seems functional.  To check, I installed the developer dependencies and ran tox, as indicated on the installation page. It got to 'congratulations :)' then had this error:

'''
_________________________________________ summary _________________________________________
  py37: commands succeeded
  py38: commands succeeded
  py39: commands succeeded
  congratulations :)
Traceback (most recent call last):
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\py\_error.py", line 66, in checked_call
    return func(*args, **kwargs)
OSError: [Errno 22] Invalid argument: 'C:\\Users\\Laurel\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\python.exe'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\Scripts\tox.exe\__main__.py", line 7, in <module>
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\session\__init__.py", line 44, in cmdline
    main(args)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\session\__init__.py", line 69, in main
    exit_code = session.runcommand()
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\session\__init__.py", line 197, in runcommand
    return self.subcommand_test()
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\session\__init__.py", line 225, in subcommand_test
    run_sequential(self.config, self.venv_dict)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\session\commands\run\sequential.py", line 14, in run_sequential
    develop_pkg(venv, config.setupdir)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\session\commands\run\sequential.py", line 28, in develop_pkg
    venv.developpkg(setupdir, action)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\venv.py", line 393, in developpkg
    self.install_pkg(setupdir, action, "develop-inst", is_develop=True)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\venv.py", line 376, in install_pkg
    self.finish()
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\venv.py", line 318, in finish
    live_config = self._getliveconfig()
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\venv.py", line 289, in _getliveconfig
    base_resolved_python_sha256 = getdigest(base_resolved_python_path)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\tox\venv.py", line 669, in getdigest
    return path.computehash("sha256")
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\py\_path\local.py", line 240, in computehash
    f = self.open('rb')
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\py\_path\local.py", line 369, in open
    return py.error.checked_call(open, self.strpath, mode)
  File "C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\lib\site-packages\py\_error.py", line 86, in checked_call
    raise cls("%s%r" % (func.__name__, args))
py.error.ENOTDIR: [Not a directory]: open('C:\\Users\\Laurel\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\\python.exe', 'rb')
'''

**Reply by rly** [reply]
This latest error might be an incompatibility between the Windows Store version of Python and anaconda. I found this [related issue](https://github.com/pypa/sampleproject/issues/136). In your anaconda prompt, enter `where python`. The path to the version of Python that is bundled with Anaconda should come first. If the WindowsApps version comes before the Anaconda version, then I think you need to change your `path` environment variable to make your Anaconda Python path is before (left of) your Windows Python path.

**Reply by laurelrr** [reply]
No, it looks ok

(pynwb) C:\Users\Laurel\Documents\GitHub\pynwb>where python
C:\Users\Laurel\AppData\Local\Continuum\anaconda3\envs\pynwb\python.exe
C:\Users\Laurel\AppData\Local\Microsoft\WindowsApps\python.exe

**Reply by rly** [reply]
Did you happen to install Python from the Windows/Microsoft Store? In looking up this issue, I have found several reports of issues with that version. e.g., https://dev.to/naruaika/why-i-didn-t-install-python-from-the-microsoft-store-5cbd 

If you did, then I would recommend uninstalling the version of Python that came from Windows/Microsoft. Then `tox` and `anaconda` will not get confused and try to use the one from Windows, which is what the path `'C:\Users\Laurel\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe'` in the error message refers to. If you want to reinstall Python outside of Anaconda, I would recommend installing Python directly from the [Python website](https://www.python.org/downloads/).

**Reply by laurelrr** [reply]
I don't recall specifically installing python from the Microsoft store.  I installed Anaconda when I first got the machine,  around 3 years ago.  Since then, I usually set up a conda environment for a specific python version.  I also have WSL installed on this machine and that does have various python versions, which I installed through the command line in ubuntu, usually using miniconda.

**Reply by rly** [reply]
That's odd. Since you said that you are OK with this for the time being, do you want to continue troubleshooting? If so, then I think the next step would be to uninstall all non-Anaconda, non-WSL versions of Python on your machine: Go to Settings > Apps and Features > scroll down and select Python > click Uninstall. Then, in a conda environment with the developer dependencies installed, please try running `tox` again. If there is still an error, please report here your PATH environment variable.
